avm/src/vm/memory.rs

```diff
@@ -1,654 +1,390 @@
-    }
-    // first check if the data is stored in a pointer
-    if self.registers_ish[addr as usize] > 0 {
-      return if self.registers_ish[addr as usize] == HM_PTR_ADDR {
-        let var = self.read_fractal_mem(addr);
-        (var, 0)
-      } else {
-        (vec![self.read_fixed(addr)], 8)
-      }
-    }
-    // test if the data read is itself a string/array
-    return if self.either_mem[addr as usize] > -1 {
-      let var = self.read_fractal_mem(addr);
-      (var, 0)
-    } else {
-      // Nope, it's fixed data. We can safely read 8 bytes for all of the fixed types
-      (vec![self.read_fixed(addr)], 8)
-    };
+  pub fn register(self: &mut HandlerMemory, addr: i64, orig_addr: i64) {
+    let (a, b) = self.addr_to_idxs(orig_addr);
+    self.set_addr(addr, a, b);
   }
 
-  pub fn read_fixed(self: &HandlerMemory, addr: i64) -> i64 {
-    if addr < 0 {
-      // global memory
-      if self.is_neg_addr_gmem(addr) {
-        let a = (0 - addr - 1) as usize;
-        let result = i64::from_ne_bytes((&self.gmem[a..a+8]).try_into().unwrap());
-        return result;
-      }
-      // closure arguments memory
-      let closure = &self.closure_args[0];
-      return closure.read_fixed(addr - CLOSURE_ARG_MEM_START);
-    }
-    let reg = self.registers_ish[addr as usize];
-    if reg == NORMAL_ADDR {
-      unsafe {
-        return self.mem.as_ptr().add(addr as usize).read();
-      }
-    } else if reg == FX_PTR_ADDR {
-      unsafe {
-        let ptr = usize::from_ne_bytes(self.mem[addr as usize].to_ne_bytes()) as *const i64;
-        let val = *ptr;
-        return val;
-      }
-    } else {
-      panic!("Trying to get fixed value from fractal pointer");
-    }
+  pub fn register_in(self: &mut HandlerMemory, orig_addr: i64, fractal_addr: i64, offset_addr: i64) {
+    let (a, b) = self.addr_to_idxs(orig_addr);
+    let mem = self.read_mut_fractal(fractal_addr);
+    mem[offset_addr as usize] = (a, b as i64);
   }
 
-  // returns a copy
-  pub fn read_fractal_mem(self: &HandlerMemory, addr: i64) -> Vec<i64> {
-    if addr < 0 {
-      // global memory
-      if self.is_neg_addr_gmem(addr) {
-        let a = (0 - addr - 1) as usize;
-        let result = &self.gmem[a..];
-        let mut out: Vec<i64> = Vec::new();
-        for i in 0..(result.len() / 8) {
-          let num = i64::from_ne_bytes((&result[8*i..8*i+8]).try_into().unwrap());
-          out.push(num);
-        }
-        return out;
-      }
-      // closure arguments memory
-      let closure = &self.closure_args[0];
-      return closure.read_fractal_mem(addr - CLOSURE_ARG_MEM_START);
-    }
-    let reg = self.registers_ish[addr as usize];
-    if reg == NORMAL_ADDR {
-      let a = addr as usize;
-      let arr = &self.fractal_mem[self.either_mem[a] as usize];
-      let res = arr.mem.as_slice();
-      return res.to_vec();
+  pub fn register_out(self: &mut HandlerMemory, fractal_addr: i64, offset_addr: i64, out_addr: i64) {
```

Comment:
nit - register_to_fractal or a quick one liner on the semantics of register_out?

Reply:
I'll add comments for all three.